Final pull request for 5.8 Write API extensions

GRyall opened the pull request:
Depends on ~#157, #161, #177, #179, #180, #187, #189, #193, #196, #208, #220~ 

Adds remaining write API functionality and functional tests for all write API methods.

See also: https://wiki.egi.eu/wiki/GOCDB/Write_API/Technical_Documentation#Supported_methods_in_GOCDB_5.8 (documentation will be merged following release)

ineilson commented:
It may be a silly question but what issue does this PR address? Where's the brief?

gregcorbett commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -902,14 +904,132 @@ private function updateEndpoint (Site $siteService) {
+   */
+  private function updateEndpointPropPost (\EndpointLocation $endpoint, $endpointPropName, $endpointPropName) {
+    if($endpointPropName == 'monitored') {
+      #Not valid for boolians
```
Why isn't `POST` valid for booleans?

GRyall replied:
We don't have any null booleans - they are already defined as true or false. Post shouldn't overwrite existing values.

gregcorbett replied:
```suggestion
      #Not valid for Endpoint booleans as they are set when entity created.
```

GRyall replied:
review: for all comments update

gregcorbett commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1798,51 +1798,69 @@ public function editEndpoint(\User $user, \EndpointLocation $endpoint, $newValue
-        if ($newValues ['SERVICEENDPOINT'] ['INTERFACENAME'] != '') {
-            $interfaceName = $newValues ['SERVICEENDPOINT'] ['INTERFACENAME'];
-        } else {
-            $interfaceName = ( string ) $service->getServiceType ();
```
Do we no longer need to catch this case?

gregcorbett commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -902,14 +904,132 @@ private function updateEndpoint (Site $siteService) {
     }
+
+    // check endpoint's name is unique under the service
+    foreach ( $endpoint->getService()->getEndpointLocations () as $endpointL ) {
```
This is done twice?

GRyall replied:
Yep. In the service service it is for the web portal as well. Doing it here ensures an API appropriate error message.

GRyall replied:
(and HTTP response code)

gregcorbett commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1798,51 +1798,69 @@ public function editEndpoint(\User $user, \EndpointLocation $endpoint, $newValue
+      }
+
+      // check endpoint's name is unique under the service
+      foreach ( $service->getEndpointLocations () as $endpointL ) {
```
This is done twice?

GRyall replied:
Yep. Here it is for the web portal as well. There it ensures an error message (and response code) on the API that is appropriate.

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -902,14 +904,132 @@ private function updateEndpoint (Site $siteService) {
+  }
+
+  /**
+   * Update a property of an endpoint folloeing a PUT requests
```
```suggestion
   * Update a property of an endpoint following a PUT request
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -902,14 +904,132 @@ private function updateEndpoint (Site $siteService) {
+  }
+
+  /**
+   * Update a property of an endpoint folloeing a POST requests
```
```suggestion
   * Update a property of an endpoint following a POST request
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -902,14 +904,132 @@ private function updateEndpoint (Site $siteService) {
+      }
     }
+
+    // check endpoint's name is unique under the service
```
```suggestion
    // check endpoint name is unique under the service
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1876,6 +1894,50 @@ public function deleteEndpoint(\EndpointLocation $endpoint, \User $user) {
     }
 
+    /**
+     * Function to establish if an endpoint already has a value sets
```
```suggestion
     * Function to establish if an endpoint already has a value set
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -776,6 +805,137 @@ private function updateService(Site $siteService){
   }
 
+  /**
+   * Update a property of a service folloeing a POST requests
```
```suggestion
   * Update a property of a service following a POST request
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -776,6 +805,137 @@ private function updateService(Site $siteService){
+    $serviceBoolPropNames = array('production', 'beta', 'monitored', 'notify');
+
+    if(in_array($servicePropName, $serviceBoolPropNames)) {
+      #Not valid for boolians
```
```suggestion
      #Not valid for booleans
```

GRyall commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1798,51 +1798,69 @@ public function editEndpoint(\User $user, \EndpointLocation $endpoint, $newValue
+          if ($endpoint != $endpointL && $endpointL->getName () == $name) {
+              throw new \Exception ( "Please provide a unique name for this endpoint." );
+          }
+      }
```
Review: abstract this out as function as reused elsewhere

GRyall replied:
see latter commit!

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -776,6 +805,137 @@ private function updateService(Site $siteService){
+  }
+
+  /**
+   * Update a property of a service folloeing a PUT requests
```
```suggestion
   * Update a property of a service following a PUT request
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -776,6 +805,137 @@ private function updateService(Site $siteService){
+   * @throws \Exception
+   */
+  private function updateServicePropPut (\Service $service, $servicePropName, $servicePropValue) {
+    #update the values
```
Comment seems redundant.
```suggestion
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -577,103 +577,153 @@ public function editService(\Service $se, $newValues, \User $user = null) {
+     * @throws Exception
+     */
+    private function editServiceLogic(\Service $service, $scopes, $sType, $updatedServiceValues){
+      // Explicity demarcate our tx boundary
```
```suggestion
      // Explicitly demarcate our tx boundary
```
What is `tx`?

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -577,103 +577,153 @@ public function editService(\Service $se, $newValues, \User $user = null) {
+
+          $service->setServiceType($sType);
+
+          // Update the service's scope
```
```suggestion
          // Update the scope of the service
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -577,103 +577,153 @@ public function editService(\Service $se, $newValues, \User $user = null) {
-            foreach ( $selectedScopesToApply as $scope ) {
-                $se->addScope ( $scope );
-            }
+          // find then link each scope specified to the site
```
I'm having trouble parsing this, is it:
"find then link each scope (specified to the site)"
or
"find then link (each scope specified) to the site"
?

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1703,6 +1753,70 @@ public function moveService(\Service $Service, \Site $Site, \User $user = null)
     }
 
+    /**
+     * Function to establish if a service already has a value sets
```
```suggestion
     * Function to establish if a service already has a value set
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -278,19 +278,20 @@ private function validateEntityTypePropertyAndPropValue() {
 
         #Now we itterate through all the possible cases.
-        #The first is that a entity key has been provided, currently only extension properties support this
+        #The first is that a entity key has been provided as well as a value, currently only extension property support this
```
```suggestion
        #The first is that a entity key has been provided as well as a value, currently only extension properties support this
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -278,19 +278,20 @@ private function validateEntityTypePropertyAndPropValue() {
         #The second is that there is no key, and a single value has been provided
         #The third is that a series of key/value pairs have been provided
-        #The fourth is a delete where no array or value has been specified, this is only currently supported for extension properties
+        #The fourth is a delete where no array or value has been specified, but a key has. This is only currently supported for deletion of xtension properties and endpoints
```
```suggestion
        #The fourth is a delete where no array or value has been specified, but a key has. This is only currently supported for deletion of extension properties and endpoints
```

tofu-rocketry replied:
```suggestion
        #The fourth is a delete where no array or value has been specified, but a key has. This is only currently supported for deletion of extension properties and endpoints
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -314,18 +315,34 @@ private function validateEntityTypePropertyAndPropValue() {
-                      "The API currently only supports deleting without specifying values for extension properties. For help see: $this->docsURL"
-                    );
+              if ($this->entityType == 'service' && $this->entityProperty == 'endpoint') {
+                #For deletion of endpoints a key must be specified
```
Not entirely clear here that this is the ID mentioned lower down.

GRyall commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1876,6 +1894,50 @@ public function deleteEndpoint(\EndpointLocation $endpoint, \User $user) {
+        return false;
+      } else {
+        return true;
+      }
```
reverse logic for readability

gregcorbett replied:
for clarity, what I meant was replacing the `if`/`else` statement with something like `return ! empty($propValue)`

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -799,6 +816,17 @@ private function updateService(Site $siteService){
+      case 'endpoint':{
+        if ($this->requestMethod != 'DELETE') {
+          $this->exceptionWithResponseCode(400,
+            "Only deletion of endpoints is supported when a service is specified. ".
```
Should this be the following? It's not clear to me...
```suggestion
            "Deletion of endpoints is only supported when a service is specified. ".
```

GRyall replied:
I believe this has been superseded by a refactor - I can't find it in the current codebase

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1042,6 +1070,39 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
   }
 
+  /**
+   * Deletes the endpoint witht the specified id, as long as it belongs to the
```
```suggestion
   * Deletes the endpoint with the specified id, as long as it belongs to the
```

tofu-rocketry replied:
```suggestion
   * Deletes the endpoint with the specified ID, as long as it belongs to the
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1042,6 +1070,39 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
 
+  /**
+   * Deletes the endpoint witht the specified id, as long as it belongs to the
+   * specified service and the user is authroised to do so
```
```suggestion
   * specified service and the user is authorised to do so
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1042,6 +1070,39 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
+      $this->exceptionWithResponseCode(404, "An endpoint with the specified ID could not be found");
+    }
+
+    #check enspoint belongs to service specified
```
```suggestion
    #check endpoint belongs to service specified
```

GRyall replied:
sp

gregcorbett commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1703,6 +1753,70 @@ public function moveService(\Service $Service, \Site $Site, \User $user = null)
+        }
+      }
+
+      if (empty($propValue)) {
```
review: reverse logic for readability

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1042,6 +1070,39 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
+    #This case requires the serviceService
+    $this->checkServiceServiceSet();
+
+    #Identify the SE
```
```suggestion
    #Identify the ServiceEndPoint
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1042,6 +1070,39 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
+    #Authorisation
+    $this->checkAuthorisation ($siteService, $service->getParentSite(), $this->userIdentifier, $this->userIdentifierType);
+
+    #delte endpoint
```
```suggestion
    #delete endpoint
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -2083,6 +2081,23 @@ public function EndpointPropSet (\EndpointLocation $endpoint, $endpointPropName)
+    /**
+     * Returns true if an endpoint with a given name exists for a given service
+     *
+     * TODO: This could be made more effecient using a DQL select statement
```
```suggestion
     * TODO: This could be made more efficient using a SQL select statement
```

tofu-rocketry replied:
Clarify what DQL is?

GRyall replied:
Doctrine Query Language - used elsewhere in codebase

tofu-rocketry commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -314,18 +315,34 @@ private function validateEntityTypePropertyAndPropValue() {
+        } elseif (!is_null($this->entityPropertyKey) && is_null($this->entityPropertyValue) && !is_null($this->entityPropertyKVArray) && $this->requestMethod == 'DELETE') {
+          if ($this->requestMethod == 'DELETE'){
+            $this->exceptionWithResponseCode(400,
+              "When using the DELETE method and specifing both the ID of the object being deleted and the ID of the parent no request body should be provided. For help see: $this->docsURL"
```
```suggestion
              "When using the DELETE method, and specifying both the ID of the object being deleted and the ID of the parent, no request body should be provided. For help see: $this->docsURL"
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1858,10 +1858,8 @@ public function addEndpoint($values, \User $user = null) {
-                throw new \Exception ( "Please provide a unique name for this endpoint." );
-            }
+        if($this->endpointWithNameExists($service, $name)){
+          throw new \Exception ( "Please provide a unique name for this endpoint." );
```
```suggestion
          throw new \Exception ( "Please provide a unique name for this Service Endpoint." );
```
?

tofu-rocketry commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1993,19 +1993,49 @@ public function deleteEndpoint(\EndpointLocation $endpoint, \User $user) {
+     *
+     * @param  EndpointLocation $endpoint endpoint to be deleted
+     * @param  Service          $service  service of endpoint being deleted
+     * @thows Exception
```
The others use 2 spaces.

```suggestion
     * @thows  Exception
```

jrha commented on htdocs/PI/write/PIWriteRequest.php:
```diff
@@ -1070,6 +1096,167 @@ private function updateServiceExtensionPropertiesDelete ($service, $extensionPro
+  /**
+   * Function to process a PUT request for a whole endpoint
+   *
+   * @param  Service $service        the service to which the endpoint is to belong (or already belongs)
```
> the service to which the endpoint is to belong (or already belongs)

?

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1921,19 +1961,41 @@ public function editEndpoint(\User $user, \EndpointLocation $endpoint, $newValue
     }
 
+    /**
+     * Edit the given endpoint using the given values provided fromt the API
```
```suggestion
     * Edit the given endpoint using the given values provided from the API
```

jrha commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -2098,6 +2160,27 @@ public function endpointWithNameExists (\Service $service, $name) {
+    /**
+     * Function to return an endpoint for a given service with a given name
+     *
+     * TODO: This could be made more effecient using a DQL select statement
```
```suggestion
     * TODO: This could be made more efficient using a Doctrine Query Language select statement
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+use Doctrine\ORM\EntityManager;
+
+/**
+ * Abstract class for testing the Writ API.
```
```suggestion
 * Abstract class for testing the Write API.
```

gregcorbett replied:
```suggestion
 * Abstract class for testing the Write API.
```

GRyall commented on lib/Gocdb_Services/ServiceService.php:
```diff
@@ -1858,10 +1858,8 @@ public function addEndpoint($values, \User $user = null) {
-        foreach ( $service->getEndpointLocations () as $endpointL ) {
-            if ($endpointL->getName () == $name) {
-                throw new \Exception ( "Please provide a unique name for this endpoint." );
-            }
```
Check that all sets of this logic have been removed (see first commit)

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  */
+  protected function getDataSet() {
+    return $this->createFlatXMLDataSet(__DIR__ . '/../doctrine/truncateDataTables.xml');
+    # Use below to return an empty data set if we don't want to truncate and seed
+    #return new PHPUnit_Extensions_Database_DataSet_DefaultDataSet();
```
One should "use" this by un-commenting the line?
Why would this be needed?

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  */
+  protected function getSetUpOperation() {
+    # CLEAN_INSERT is default
+    #return PHPUnit_Extensions_Database_Operation_Factory::CLEAN_INSERT();
+    #return PHPUnit_Extensions_Database_Operation_Factory::UPDATE();
+    #return PHPUnit_Extensions_Database_Operation_Factory::NONE();
```
When would these be needed?

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  /**
+   * Each methods test class should test the case of unathenticated requests.
```
```suggestion
   * The test class for each method should test the case of unauthenticated requests.
```

tofu-rocketry replied:
```suggestion
   * Each method's test class should test the case of unauthenticated requests.
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  abstract protected function test_APIUnauthenticated();
+
+  /**
+   * Each methods test class should test the case of unathorised requests.
```
```suggestion
   * The test class for each method should test the case of unauthorised requests.
```

tofu-rocketry replied:
```suggestion
   * Each method's test class should test the case of unauthorised requests.
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+
+  /**
+   * Function to make a call to the API for tests that assumes we are testing
+   * functionality and not trying to break it through poorly formated requests
```
```suggestion
   * functionality and not trying to break it through poorly formatted requests
```

tofu-rocketry replied:
```suggestion
   * functionality and not trying to break it through poorly formatted requests
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  /**
+   * Create a Json string in a format acceptable to the Write API for a single value
```
```suggestion
   * Create a JSON string in a format acceptable to the Write API for a single value
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  /**
+   * Create a Json string in a format acceptable to the Write API for a single
```
```suggestion
   * Create a JSON string in a format acceptable to the Write API for a single
```

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  /**
+   * Creates a sample site for testing which is a memebr of a sample NGI.
```
```suggestion
   * Creates a sample site for testing which is a member of a sample NGI.
```

GRyall commented:
review: inconsistency in function names in write API  - need to consistently append or prefix method

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+ *
+ */
+class WriteAPIendpointMethodsTests extends extensionPropertyAbstract {
+  #Some sample data to test with. Case mixing is deliberate - the API should be case insensitive
```
If this failed due to case sensitivity magically happening, would it be obvious why?

GRyall replied:
discussed in review. No - but there isn't a simple way of testing this, so this is a quick way of ensuring it is covered

tofu-rocketry commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+  abstract protected function test_APIUnauthorised();
+
+  /**
+   * Function to make a call to the write API and return an aray containing the
```
```suggestion
   * Function to make a call to the write API and return an array containing the
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+   */
+  public function test_endpointDeletionNoRequest(){
+    print __METHOD__ . "\n";
+    #null is wht the value the request body will have in an API call when no body
```
```suggestion
    #null is what the value the request body will have in an API call when no body
```

tofu-rocketry commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+      $this->assertServiceBooleansAsExpected($service->getId(),array($field =>$value));
+    }
+
+    #And try switching the boolean values back (we have had a number of bugs
```
You spelt boolean right!

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+   *                                       if it exists (ennables testing wrong IDs)
+   */
+  private function endpointDeletion($shouldSucceed, $endpointId, $serviceId, $request, $expectedHttpResponse, $endpointIdToCheck=null){
+    #case mixuing intentional as write API should be mostly case insensitve
```
```suggestion
    #case mixing intentional as write API should be mostly case insensitive
```
mostly!?

tofu-rocketry commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+      $this->assertServiceBooleansAsExpected($service->getId(),array($field =>$value));
+    }
+
+    #Try and change all non-boolian properties again and check it is successful (which it should be for PUT)
```
```suggestion
    #Try and change all non-boolean properties again and check it is successful (which it should be for PUT)
```

jrha replied:
```suggestion
    #Try and change all non-boolean properties again and check it is successful (which it should be for PUT)
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+
+    #It should be possible to update each of the non-mandatory, non-boolean endpoint properties
+    foreach ($this->sampleValuesOptionalFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
```
```suggestion
      #case mixing intentional as write API should be mostly case insensitive
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+
+    #It should be possible to update each of the non-boolean endpoint Site_Properties
+    foreach ($this->sampleValuesOptionalFields + $this->sampleValuesMandatoryFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
```
```suggestion
      #case mixing intentional as write API should be mostly case insensitive
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+      $this->assertEndpointBooleansAsExpected($endpoint->getId(),array($field =>$value));
+    }
+
+    #Try and change all non-boolian properties again and check it is successful (which it should be for PUT)
```
```suggestion
    #Try and change all non-boolean properties again and check it is successful (which it should be for PUT)
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+  /*
+   *This function provides each instance of the class a method to feed in a strings
+   *saying ehat type of entity it is.
```
```suggestion
   *saying what type of entity it is.
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+      #As it is unauthenticated it should return a 403
+      $this->assertequals(403,$APIOutput['httpResponseCode']);
+
+      #Check that creating wasn't successful
```
```suggestion
      #Check that creating was unsuccessful
```
Minor, but I think it implies that it *should* fail, rather than it being an accident.

jrha commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+   * @param  string $entKey          optionally the key to a value of the propery being changed
+   * @return array                   array containing response code and API response
+   */
+  protected function wellFormatedWriteAPICall ($method, $requestContents, $authIdent, $entType, $entID, $entProp, $entKey = null ) {
```
```suggestion
  protected function wellFormattedWriteAPICall ($method, $requestContents, $authIdent, $entType, $entID, $entProp, $entKey = null ) {
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    $values = $valuesLessBool + $this->boolFieldsStart;
+
+    #mixed case deliberate - the API should be case insensive
+    $APIOutput = $this->wellFormatedWriteAPICall ($method, json_encode($values), $this->validAuthIdent, 'servIce', $service->getId(), 'enDpoint');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ($method, json_encode($values), $this->validAuthIdent, 'servIce', $service->getId(), 'enDpoint');
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    $newValues = $newValuesLessBool + $this->boolFields;
+
+
+    $APIOutput = $this->wellFormatedWriteAPICall ($method, json_encode($newValues), $this->validAuthIdent, 'service', $service->getId(), 'endpoint');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ($method, json_encode($newValues), $this->validAuthIdent, 'service', $service->getId(), 'endpoint');
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+   */
+  private function endpointDeletion($shouldSucceed, $endpointId, $serviceId, $request, $expectedHttpResponse, $endpointIdToCheck=null){
+    #case mixuing intentional as write API should be mostly case insensitve
+    $APIOutput = $this->wellFormatedWriteAPICall ('DelEte', $request, $this->validAuthIdent, 'ServICe', $serviceId, 'EndPOint', $endpointId);
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DelEte', $request, $this->validAuthIdent, 'ServICe', $serviceId, 'EndPOint', $endpointId);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #It should be possible to update each of the non-mandatory, non-boolean endpoint properties
+    foreach ($this->sampleValuesOptionalFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
+      $APIOutput = $this->wellFormatedWriteAPICall ('PoST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endPoiNt', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PoST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endPoiNt', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+
+    #Try and change each field
+    foreach ($changes as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #Now try putting with some boolean properties. These should fail for POST
+    #Try and change each field
+    foreach ($this->boolFields as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #they are unchanged later
+    $startingMandFileds =  array('name' => $endpoint->getname(), 'url'=>$endpoint->getUrl());
+
+    $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'endpoint', $endpoint->getId(), 'notARealField');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'endpoint', $endpoint->getId(), 'notARealField');
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #It should be possible to update each of the non-boolean endpoint Site_Properties
+    foreach ($this->sampleValuesOptionalFields + $this->sampleValuesMandatoryFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUt', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endPOint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUt', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endPOint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+
+    #It should also be possible to update each of the booleans endpoint properties
+    foreach ($this->boolFields as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #And try switching the boolean values back (we have had a number of bugs
+    #where switching in one direction works but not the other)
+    foreach ($this->boolFieldsStart as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+
+    #Try and change each field
+    foreach ($changes as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'endpoint', $endpoint->getId(), $field);
```

jrha commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+    #they are unchanged later
+    $startingMandFileds =  array('name' => $endpoint->getName(), 'url'=>$endpoint->getUrl());
+
+    $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'endpoint', $endpoint->getId(), 'notARealField');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'endpoint', $endpoint->getId(), 'notARealField');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+    foreach (array('POST','PUT','DELETE') as $method) {
+      #The write API index uses null for the authentication identifier if the user is unauthenticated
+      $APIOutput = $this->wellFormatedWriteAPICall ($method, json_encode($this->sampleExtensionProps), null, $entType, $entity->getId(), 'extensionproperties');
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ($method, json_encode($this->sampleExtensionProps), null, $entType, $entity->getId(), 'extensionproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $entType = $this->getSampleEntityType();
+
+    foreach (array('POST','PUT','DELETE') as $method) {
+      $APIOutput = $this->wellFormatedWriteAPICall ($method, json_encode($this->sampleExtensionProps), 'definitly not an authenticated user', $entType, $entity->getId(), 'extensionproperties');
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ($method, json_encode($this->sampleExtensionProps), 'definitly not an authenticated user', $entType, $entity->getId(), 'extensionproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $entType = $this->getSampleEntityType();
+
+    #case variation in method and key is deliberate as API should be insensitive
+    $APIOutputService = $this->wellFormatedWriteAPICall ('PoSt', json_encode($this->sampleExtensionProps), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('PoSt', json_encode($this->sampleExtensionProps), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+
+    #Now try adding a new set of props, one of which has the same key as the previous
+    $APIOutputService = $this->wellFormatedWriteAPICall ('POST', json_encode($this->sampleExtensionPropsWDup), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties');
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('POST', json_encode($this->sampleExtensionPropsWDup), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $entType = $this->getSampleEntityType();
+
+    #case variation in method and key is deliberate as API should be insensitive
+    $APIOutputService = $this->wellFormatedWriteAPICall ('Put', json_encode($this->sampleExtensionProps), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('Put', json_encode($this->sampleExtensionProps), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+
+    #Now try adding a new set of props, one of which has the same key as the previous
+    $APIOutputService = $this->wellFormatedWriteAPICall ('PUT', json_encode($this->sampleExtensionPropsWDup), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties');
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('PUT', json_encode($this->sampleExtensionPropsWDup), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $entType = $this->getSampleEntityType();
+
+    #case variation in method and key is deliberate as API should be insensitive
+    $APIOutputService = $this->wellFormatedWriteAPICall ('PoSt', $this->singleValToJsonRequest($this->sampleExtensionPropVal), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $this->sampleExtensionPropKey );
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('PoSt', $this->singleValToJsonRequest($this->sampleExtensionPropVal), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $this->sampleExtensionPropKey );
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+
+    #Now try adding a new property where the key is the same as the one we allready added
+    $APIOutputService = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest($this->sampleExtensionPropNewVal), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties', $this->sampleExtensionPropKey );
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest($this->sampleExtensionPropNewVal), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties', $this->sampleExtensionPropKey );
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $entType = $this->getSampleEntityType();
+
+    #case variation in method and key is deliberate as API should be insensitive
+    $APIOutputService = $this->wellFormatedWriteAPICall ('PUt', $this->singleValToJsonRequest($this->sampleExtensionPropVal), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $this->sampleExtensionPropKey );
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('PUt', $this->singleValToJsonRequest($this->sampleExtensionPropVal), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $this->sampleExtensionPropKey );
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+
+    #Now try adding a new property where the key is the same as the one we allready added
+    $APIOutputService = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($this->sampleExtensionPropNewVal), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties', $this->sampleExtensionPropKey );
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($this->sampleExtensionPropNewVal), $this->validAuthIdent, $entType, $ent->getId(), 'extensionproperties', $this->sampleExtensionPropKey );
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    }
+
+    #case variation in method and key is deliberate as API should be insensitive
+    $APIOutputService = $this->wellFormatedWriteAPICall ('DEleTE', $requestContents, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutputService = $this->wellFormattedWriteAPICall ('DEleTE', $requestContents, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    }
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $propsToDelete[array_keys(array_slice($this->sampleExtensionProps, 0, 1))[0] . 'Not the same now'] = array_values(array_slice($this->sampleExtensionProps, 0, 1))[0];
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $propsToDelete[array_keys(array_slice($this->sampleExtensionProps, 0, 1))[0]] = array_values(array_slice($this->sampleExtensionProps, 0, 1))[0].'Not the Same';
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $propKey = array_keys(array_slice($this->sampleExtensionProps, 0, 1))[0];
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties',$propKey);
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties',$propKey);
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $propKey = array_keys(array_slice($this->sampleExtensionProps, 0, 1))[0] . 'this doesnt exist';
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties',$propKey);
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties',$propKey);
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    $propKey = array_keys(array_slice($this->sampleExtensionProps, 0, 1))[0];
+
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $propKey);
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $propKey);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #It should be possible to update each of the non-mandatory, non-boolean service Site_Properties
+    foreach ($this->sampleValuesOptionalFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
+      $APIOutput = $this->wellFormatedWriteAPICall ('PoST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'serVice', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PoST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'serVice', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+
+    #Try and change each field
+    foreach ($changes as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #Now try putting with some boolean properties. These should fail for POST
+    #Try and change each field
+    foreach ($this->boolFields as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #they are unchanged later
+    $startingMandFileds =  array('hostname' => $service->getHostname(), 'email'=>$service->getEmail());
+
+    $APIOutput = $this->wellFormatedWriteAPICall ('POST', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'service', $service->getId(), 'notARealField');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('POST', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'service', $service->getId(), 'notARealField');
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #It should be possible to update each of the non-boolean service Site_Properties
+    foreach ($this->sampleValuesOptionalFields + $this->sampleValuesMandatoryFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUt', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'serVice', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUt', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'serVice', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+
+    #It should also be possible to update each of the booleans service properties
+    foreach ($this->boolFields as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #And try switching the boolean values back (we have had a number of bugs
+    #where switching in one direction works but not the other)
+    foreach ($this->boolFieldsStart as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+
+    #Try and change each field
+    foreach ($changes as $field => $value) {
+      $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```
```suggestion
      $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest($value), $this->validAuthIdent, 'service', $service->getId(), $field);
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+    #they are unchanged later
+    $startingMandFileds =  array('hostname' => $service->getHostname(), 'email'=>$service->getEmail());
+
+    $APIOutput = $this->wellFormatedWriteAPICall ('PUT', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'service', $service->getId(), 'notARealField');
```
```suggestion
    $APIOutput = $this->wellFormattedWriteAPICall ('PUT', $this->singleValToJsonRequest('string'), $this->validAuthIdent, 'service', $service->getId(), 'notARealField');
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+
+    #It should be possible to update each of the non-boolean service Site_Properties
+    foreach ($this->sampleValuesOptionalFields + $this->sampleValuesMandatoryFields as $field => $value) {
+      #case mixuing intentional as write API should be mostly case insensitve
```
```suggestion
      #case mixing intentional as write API should be mostly case insensitive
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+  /**
+   * The fieldnames used for services by the GOCDB schema, and so by the API, do
+   * not match the names we use internally for or DB fields in all cases. This
+   * function takes the name used in the API and if there is a known missmatch,
```
```suggestion
   * function takes the name used in the API and if there is a known mismatch,
```

jrha commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+   * The fieldnames used for services by the GOCDB schema, and so by the API, do
+   * not match the names we use internally for or DB fields in all cases. This
+   * function takes the name used in the API and if there is a known missmatch,
+   * corrects it. The db fields are also all lower case, so we corrct for that as well.
```
```suggestion
   * corrects it. The db fields are also all lower case, so we correct for that as well.
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+  /**
+   * Assert that the given entity has no extension properties. As the property
+   * values are implmented almost identically for the three entities that have
```
```suggestion
   * values are implemented almost identically for the three entities that have
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+   */
+  protected function assertNoEntityProperties($entityType, $entityId) {
+    #Depending which entity we are looking at the name of the parent ID field
+    #is differnt and we will want to look at a differnt table in each case
```
```suggestion
    #is different and we will want to look at a different table in each case
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    }
+
+    #Depending which entity we are looking at the name of the parent ID field
+    #is differnt and we will want to look at a differnt table in each case
```
```suggestion
    #is different and we will want to look at a different table in each case
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+  /**
+   * Depending which entity we are looking at the name of the extension property
+   * table is differnt. This function returns the correct name
```
```suggestion
   * table is different. This function returns the correct name
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties', $propKey);
+
+    #It was a properly formated request, but one of the properties values did not
```
```suggestion
    #It was a properly formatted request, but one of the properties values did not
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+  }
+
+  /**
+   * Use the API to delete a single extension property, but sepcify the property
+   * using the wrong value witht he correct key. This should fail.
```
```suggestion
   * Use the API to delete a single extension property, but specify the property
   * using the wrong value with the correct key. This should fail.
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+      $reqBody = '';
+    } else {
+      #(null replicates the output of file_get_contents('php://input') when
+      #(there is no request contents)
```
```suggestion
      #(there are no request contents)
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+
+  /**
+   * Use the API to delete a single extension property for the specified entity
+   * that doesnt exist. Checks that the correct errors are thrown
```
```suggestion
   * that doesn't exist. Checks that the correct errors are thrown
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', $reqBody, $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties',$propKey);
+
+    #It was a properly formated request, so the API should return a response code of 204 (for now)
```
```suggestion
    #It was a properly formatted request, so the API should return a response code of 204 (for now)
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+      $reqBody = '';
+    } else {
+      #(null replicates the output of file_get_contents('php://input') when
+      #(there is no request contents)
```
```suggestion
      #(there are no request contents)
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
+
+    #It was a properly formated request, but one of the properties values did not
```
```suggestion
    #It was a properly formatted request, but one of the properties values did not
```

jrha commented on tests/writeAPI/extensionPropertyAbstract.php:
```diff
@@ -0,0 +1,710 @@
+    #Use the API. (case variation in method and key is deliberate as API should be insensitive)
+    $APIOutput = $this->wellFormatedWriteAPICall ('DEleTE', json_encode($propsToDelete), $this->validAuthIdent, $entType, $ent->getId(), 'EXTENSIONproperties');
+
+    #It was a properly formated request, but one of the properties does not exist,
```
```suggestion
    #It was a properly formatted request, but one of the properties does not exist,
```

tofu-rocketry commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+   * functionality and not trying to break it through poorly formated requests
+   * @param  string $method          HTTP method of request
+   * @param  string $requestContents HTTP request contents
+   * @param string $authIdent        Authentication identifier
```
```suggestion
   * @param  string $authIdent        Authentication identifier
```

gregcorbett commented on tests/writeAPI/abstractClass.php:
```diff
@@ -0,0 +1,286 @@
+
+  /**
+   * Create a Json string in a format acceptable to the Write API for a single
+   * value which is  boolean
```
```suggestion
   * value which is boolean
```

jrha commented on tests/DoctrineTestSuite1.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+ini_set('memory_limit', '512M');
```
```suggestion
// Most of our environments (including Travis) default to (or are configured to) a memory limit of 128M.
// The latest additions to the test suite nudge it over this limit.
// So we increase the limit when running the test suite to 512M.
ini_set('memory_limit', '512M');
```

gregcorbett commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+                                                    'host_os'=>'best OS Updated',
+                                                    'host_arch'=> 'archEtecture. Updated',
+                                                    'url'=>'https://some.urlupdated');
+  #The order here matters!
```
Why?

tofu-rocketry commented on tests/writeAPI/serviceMethods.php:
```diff
@@ -0,0 +1,289 @@
+  }
+
+  /**
+   * Assertthe boolean propertues of the service are as expected
```
```suggestion
   * Assert the boolean properties of the service are as expected
```

tofu-rocketry commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+ require_once __DIR__ . '/extensionPropertyAbstract.php';
+
+/**
+ * DBUnit test class for endpo methods of Write API
```
```suggestion
 * DBUnit test class for endpoint methods of Write API
```

tofu-rocketry commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+   */
+  public function test_endpointDeletionWrongEndpointIdNoRquest(){
+    print __METHOD__ . "\n";
+    #null is wht the value the request body will have in an API call when no body
```
```suggestion
    #null is what the value the request body will have in an API call when no body
```

tofu-rocketry commented on tests/writeAPI/endpointMethods.php:
```diff
@@ -0,0 +1,518 @@
+   }
+
+  /**
+   * Test the deltion of an endpoint
```
```suggestion
   * Test the deletion of an endpoint
```

GRyall commented:
I believe I have now addressed all the issues raised with this PR, with the exception of the following (which I will start working on now):

> 
> review: inconsistency in function names in write API - need to consistently append or prefix method

GRyall commented:
All comments from review addressed and merged into previous commits as agreed. All good for a final checking over and merge!

gregcorbett commented:
When running this code, I have trouble adding an `Endpoint` via the webportal. 

When I click "Add Service Endpoint" on a `Add_Service_Endpoint` page, I get a white screen of death and the following error in the logs.

```
Call to a member function getID() on a non-object in /usr/share/GOCDB5/htdocs/web_portal/controllers/service/add_service_endpoint.php on line 55
```

The endpoint does get created, I just don't get the pretty "congrat's you've created an endpoint" page.

I don't see this when running `dev`, which this branch builds directly on top of.

@GRyall , do you see the same thing?

GRyall commented:
> 
> 
> When running this code, I have trouble adding an `Endpoint` via the webportal.
> 
> When I click "Add Service Endpoint" on a `Add_Service_Endpoint` page, I get a white screen of death and the following error in the logs.
> 
> ```
> Call to a member function getID() on a non-object in /usr/share/GOCDB5/htdocs/web_portal/controllers/service/add_service_endpoint.php on line 55
> ```
> 
> The endpoint does get created, I just don't get the pretty "congrat's you've created an endpoint" page.
> 
> I don't see this when running `dev`, which this branch builds directly on top of.
> 
> @GRyall , do you see the same thing?

Looks like I missed a return statement when I refactored some code. Now fixed

GRyall commented:
> 
> 
> > When running this code, I have trouble adding an `Endpoint` via the webportal.
> > When I click "Add Service Endpoint" on a `Add_Service_Endpoint` page, I get a white screen of death and the following error in the logs.
> > ```
> > Call to a member function getID() on a non-object in /usr/share/GOCDB5/htdocs/web_portal/controllers/service/add_service_endpoint.php on line 55
> > ```
> > 
> > 
> > The endpoint does get created, I just don't get the pretty "congrat's you've created an endpoint" page.
> > I don't see this when running `dev`, which this branch builds directly on top of.
> > @GRyall , do you see the same thing?
> 
> Looks like I missed a return statement when I refactored some code. Now fixed

Actually fixed this time

gregcorbett commented:
If anyone else would like to review, please do so before lunch time Fri 14th at the latest.
